### PR TITLE
copr: Add Non-Vectorize Built-in Function `FromDays`

### PR DESCRIPTION
**Pull request by wangwangwar**
PCP #5751

Signed-off-by: wangwangwar <wangwangwar@gmail.com>

### What have you changed?

1. Add `Time` utility function `from_days`
2. Add non-vectorized built-in function `FromDays`

### What is the type of the changes?

Improvement

### How is the PR tested?

Unit test

### Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No. It should not change the behavior.

### Does this PR affect tidb-ansible?

No

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/copr-test/pull/52

###  Any examples? (optional)

### CODE REVIEW

**Comment by wangwangwar** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/52

**Comment by wangwangwar** [comment]
The copr-test here passes, but when it runs locally or in https://github.com/tikv/copr-test/pull/52, it fails with such errors:

```
[2019-11-17T05:38:50.406Z] 2019/11/17 13:38:50 2019/11/17 13:38:50 Test summary: non-matching queries: 0, success queries: 823, skipped queries: 177

[2019-11-17T05:38:50.406Z] 2019/11/17 13:38:50 Test summary(sql/randgen/6_date_1.sql): Test case PASS
... 

[2019-11-17T05:39:18.650Z] 2019/11/17 13:39:18 2019/11/17 13:39:12 Test fail: Outputs are not matching.

[2019-11-17T05:39:18.650Z] Test case: sql/randgen/6_date_2.sql

[2019-11-17T05:39:18.650Z] Statement: #1863 -  SELECT 'bzgbqflwhwcaeuxfahqfwticxkazzutgowlvzlgzzdfpkalnzpscpzhbistlpczclgtrgompddizklbmpjazufrnmuhecjevbcsgvuqkowucdsmgirehorvagwzqjvrtskxsmxwtnowiwvpancg' % `col_decimal_unsigned` AS field1, - '2014-10-23 09:31:35.063505' AS field2, TO_SECONDS( `col_time` ) AS field3 FROM `table0_int_autoinc` WHERE FROM_DAYS( `pk` ) /* QNO 1865 CON_ID 208 */ ;

[2019-11-17T05:39:18.650Z] NoPushDown Output: 

[2019-11-17T05:39:18.650Z] field1	field2	field3

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] PushDownWithoutVec Output: 

[2019-11-17T05:39:18.650Z] Error 1105: other error: Unknown signature: FromDays

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] PushDownWithVec Output: 

[2019-11-17T05:39:18.650Z] Error 1105: other error: Unknown signature: FromDays

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 2019/11/17 13:39:15 Test fail: Outputs are not matching.

[2019-11-17T05:39:18.650Z] Test case: sql/randgen/6_date_2.sql

[2019-11-17T05:39:18.650Z] Statement: #2443 -  SELECT - 12580 AS field1 FROM `table20_int_autoinc` WHERE FROM_DAYS( `col_smallint` ) /* QNO 2445 CON_ID 208 */ ;

[2019-11-17T05:39:18.650Z] NoPushDown Output: 

[2019-11-17T05:39:18.650Z] field1

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] -12580

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] PushDownWithoutVec Output: 

[2019-11-17T05:39:18.650Z] Error 1105: other error: Unknown signature: FromDays

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] PushDownWithVec Output: 

[2019-11-17T05:39:18.650Z] Error 1105: other error: Unknown signature: FromDays

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 

[2019-11-17T05:39:18.650Z] 2019/11/17 13:39:18 Test summary: non-matching queries: 2, success queries: 2461, skipped queries: 539

[2019-11-17T05:39:18.650Z] 2019/11/17 13:39:18 Test summary(sql/randgen/6_date_2.sql): Test case FAIL
```
I don't know what happened.

**Comment by wangwangwar** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/52

**Comment by wangwangwar** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/52

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 1]
```diff
@@ -1358,6 +1358,76 @@ impl Time {
...
+            fsp: 0,
+            time_type: TimeType::Date,
+        };
+        Time::new(ctx, time_args)
+    }
+
+    // Changes a daynr to year, month and day, daynr 0 is returned as date 00.00.00
+    #[inline]
+    fn get_date_from_daynr(daynr: u32) -> (u32, u32, u32) {
+        if daynr <= 365 || daynr >= 3_652_500 {
```
Comment (iosmanthus) [inline components/tidb_query/src/codec/mysql/time/mod.rs]:
```sql
mysql iosmanthus@(none):test> select from_days(3652424)                                                                           
+--------------------+
| from_days(3652424) |
+--------------------+
| 9999-12-31         |
+--------------------+
1 row in set
Time: 0.019s
mysql iosmanthus@(none):test> select from_days(3652425)                                                                           
+--------------------+
| from_days(3652425) |
+--------------------+
| 0000-00-00         |
+--------------------+
1 row in set
Time: 0.021s
```

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 2]
```diff
@@ -1358,6 +1358,76 @@ impl Time {
...
+            if day_of_year > 31 + 28 {
+                day_of_year -= 1;
+                if day_of_year == 31 + 28 {
+                    // Handle leapyears leapday.
+                    leap_day = 1;
+                }
+            }
+        }
+
+        let days_in_month: Vec<u32> = vec![31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
```
[the remark this answers is not in the document]
Comment (iosmanthus) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: just like:
```rust
    let (mut year, mut month) = (1, 1);
    days -= 365;
    for each_year in 1.. {
        let year_days = year_days(each_year);
        if days > year_days {
            year += 1;
            days -= year_days;
        } else {
            break;
        }
    }
    for each_month in 1..=12 {
        let last_day_of_month = last_day_of_month(year, each_month);
        if days > last_day_of_month {
            month += 1;
            days -= last_day_of_month;
        } else {
            break;
        }
    }

    let day = days;

    (year, month, day)

```

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 3]
```diff
@@ -1358,6 +1358,76 @@ impl Time {
...
+
+        let day = day_of_year + leap_day;
+
+        (year, month, day)
+    }
+
+    // Calculates days in one year, it works with 0 <= year <= 99.
+    #[inline]
+    fn calc_days_in_year(year: u32) -> u32 {
+        if (year & 3) == 0 && (year % 100 != 0 || (year % 400 == 0 && (year != 0))) {
```
[the remark this answers is not in the document]
Comment (iosmanthus) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: Or maybe we could just remove this function.

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 1 above)
Comment (wangwangwar) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: The implementation is copied from TiDB's, and MySQL 5.7's source code writes like this also:
```
// sql_time.cc

void get_date_from_daynr(long daynr, uint *ret_year, uint *ret_month, uint *ret_day)
{
...
    if (daynr <= 365L || daynr >= 3652500)
    {
        ...
    }
...
}
```
 Why isn't it exactly 3652424L, and it also confused me.

**Comment by iosmanthus** [comment]
LGTM! Well done!

**Comment by wangwangwar** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/52

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 4]
```diff
@@ -216,6 +216,12 @@ impl From<TimeType> for FieldTypeTp {
     }
 }
 
+impl Default for TimeType {
+    fn default() -> Self {
+        TimeType::Date
```
Comment (breeswish) [inline components/tidb_query/src/codec/mysql/time/mod.rs]: Although I know it is for deriving Default for TimeArgs, but this default itself does not make sense..

File: components/tidb_query/src/codec/mysql/time/mod.rs [hunk 5]
```diff
@@ -1358,6 +1364,50 @@ impl Time {
...
+            month,
+            day,
+            ..Default::default()
+        };
+        Time::new(ctx, time_args)
+    }
+
+    // Changes a daynr to year, month and day, daynr 0 is returned as date 00.00.00
+    #[inline]
+    fn get_date_from_daynr(daynr: u32) -> (u32, u32, u32) {
```
Comment (breeswish) [inline components/tidb_query/src/codec/mysql/time/mod.rs]: Interesting implementation. How is the performance compared to TiDB's implementation?

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 4 above)
Comment (wangwangwar) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: I got it. I remove the #[derive(.., Default)] and only impl Default for `TimeArgs`:
```Rust
impl Default for TimeArgs {
    fn default() -> Self {
        TimeArgs {
            year: 0,
            month: 0,
           day: 0,
           hour: 0,
           minute: 0,
           second: 0,
           micro: 0,
           fsp: 0,
           time_type: TimeType::Date,
        }
    }
}

File: components/tidb_query/src/codec/mysql/time/mod.rs (hunk 5 above)
Comment (wangwangwar) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: Two implementations and benchmark tests:
```Rust
impl Time {
    // Changes a daynr to year, month and day, daynr 0 is returned as date 00.00.00
    #[inline]
    pub fn get_date_from_daynr(daynr: u32) -> (u32, u32, u32) {
        if daynr <= 365 || daynr >= 3_652_425 {
            return (0, 0, 0);
        }

        let (mut year, mut month, mut days) = (1, 1, daynr);
        days -= 365;
        for each_year in 1.. {
            let days_in_year = if is_leap_year(each_year) { 366 } else { 365 };
            if days > days_in_year {
                year += 1;
                days -= days_in_year;
            } else {
                break;
            }
        }
        for each_month in 1..=12 {
            let last_day_of_month = last_day_of_month(year, each_month);
            if days > last_day_of_month {
                month += 1;
                days -= last_day_of_month;
            } else {
                break;
            }
        }

        let day = days;

        (year, month, day)
    }

    #[inline]
    pub fn get_date_from_daynr2(daynr: u32) -> (u32, u32, u32) {
        if daynr <= 365 || daynr >= 3_652_425 {
            return (0, 0, 0);
        }

        let mut year = daynr * 100 / 36525;
        let temp = (((year - 1) / 100 + 1) * 3) / 4;
        let mut day_of_year = daynr - year * 365 - (year - 1) / 4 + temp;

        let mut days_in_year = if is_leap_year(year) { 366 } else { 365 };
        while day_of_year > days_in_year {
            day_of_year -= days_in_year;
            year += 1;
            days_in_year = if is_leap_year(year) { 366 } else { 365 };
        }

        let mut leap_day = 0;
        if days_in_year == 366 && day_of_year > 31 + 28 {
            day_of_year -= 1;
            if day_of_year == 31 + 28 {
                leap_day = 1;
            }
        }

        let mut month = 1;
        let days_in_month = vec![31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        for days in days_in_month {
            if day_of_year <= days {
                break;
            }
            day_of_year -= days;
            month += 1;
        }

        let days = day_of_year + leap_day;

        (year, month, days)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[bench]
    fn bench_from_days(b: &mut Bencher) {
        b.iter(|| {
            let cases = vec![
                (140, (0, 0, 0)),          // mysql FROM_DAYS returns 0000-00-00 for any day <= 365.
                (735_000, (2012, 05, 12)), // Leap year.
                (735_030, (2012, 06, 11)),
                (735_130, (2012, 09, 19)),
                (734_909, (2012, 02, 11)),
                (734_878, (2012, 01, 11)),
                (734_927, (2012, 02, 29)),
                (734_634, (2011, 05, 12)), // Non Leap year.
                (734_664, (2011, 06, 11)),
                (734_764, (2011, 09, 19)),
                (734_544, (2011, 02, 11)),
                (734_513, (2011, 01, 11)),
                (3_652_424, (9999, 12, 31)),
                (3_652_425, (0000, 00, 00)), // mysql FROM_DAYS returns 0000-00-00 for any day >= 3652425
            ];
            for (arg, exp) in cases {
                assert_eq!(Time::get_date_from_daynr(arg), exp);
            }
        });
    }

    #[bench]
    fn bench_from_days2(b: &mut Bencher) {
        b.iter(|| {
            let cases = vec![
                (140, (0, 0, 0)),          // mysql FROM_DAYS returns 0000-00-00 for any day <= 365.
                (735_000, (2012, 05, 12)), // Leap year.
                (735_030, (2012, 06, 11)),
                (735_130, (2012, 09, 19)),
                (734_909, (2012, 02, 11)),
                (734_878, (2012, 01, 11)),
                (734_927, (2012, 02, 29)),
                (734_634, (2011, 05, 12)), // Non Leap year.
                (734_664, (2011, 06, 11)),
                (734_764, (2011, 09, 19)),
                (734_544, (2011, 02, 11)),
                (734_513, (2011, 01, 11)),
                (3_652_424, (9999, 12, 31)),
                (3_652_425, (0000, 00, 00)), // mysql FROM_DAYS returns 0000-00-00 for any day >= 3652425
            ];
            for (arg, exp) in cases {
                assert_eq!(Time::get_date_from_daynr2(arg), exp);
            }
        });
    }
}
```

Benchmark outputs:
```
# cargo bench
test codec::mysql::time::tests::bench_from_days                                           ... bench:      38,366 ns/iter (+/- 2,759)
test codec::mysql::time::tests::bench_from_days2                                          ... bench:         108 ns/iter (+/- 7)
```
Actually the implementation 1 is much slower than  2 (~300x) :(

Comment (wangwangwar) [reply components/tidb_query/src/codec/mysql/time/mod.rs]: My implementation uses iteration to compute ‘year’ and ‘day_of_year’, while the TiDB’s(and MySQL’s) uses math. The latter is much faster.

**Comment by wangwangwar** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/52

**Comment by wangwangwar** [comment]
@iosmanthus PTAL, thanks

**Pull request merged**